db/migrate/20200819030609_migrate_user_topic_timers_to_bookmark_reminders.rb

```diff
@@ -0,0 +1,81 @@
+# frozen_string_literal: true
+
+class MigrateUserTopicTimersToBookmarkReminders < ActiveRecord::Migration[6.0]
+  def up
+    topic_timers_to_migrate = DB.query(<<~SQL, reminder_type: 5)
+      SELECT topic_timers.id, topic_timers.topic_id, topic_timers.user_id, execute_at,
+      posts.id AS first_post_id
+      FROM topic_timers
+      INNER JOIN topics ON topics.id = topic_timers.topic_id
+      INNER JOIN posts ON posts.topic_id = topics.id AND posts.post_number = 1
+      WHERE topics.deleted_at IS NULL
+      AND topic_timers.deleted_at IS NULL
+      AND topic_timers.status_type = :reminder_type
+    SQL
+
+    return if topic_timers_to_migrate.empty?
+
+    topic_timer_tuples = topic_timers_to_migrate.map do |tt|
+      "(#{tt.user_id}, #{tt.topic_id})"
+    end.join(", ")
+
+    existing_bookmarks = DB.query(<<~SQL)
+      SELECT bookmarks.id, reminder_at, reminder_type,
+             bookmarks.topic_id, post_id, bookmarks.user_id, posts.post_number
+      FROM bookmarks
+      INNER JOIN posts ON posts.id = bookmarks.post_id
+      WHERE (bookmarks.user_id, bookmarks.topic_id) IN (#{topic_timer_tuples})
+    SQL
+
+    new_bookmarks = []
+    topic_timers_to_migrate.each do |tt|
+      bookmark = existing_bookmarks.find do |bm|
+
+        # we only care about existing topic-level bookmarks here
+        # because topic timers are (funnily enough) topic-level
+        bm.topic_id == tt.topic_id && bm.user_id == tt.user_id && bm.post_number == 1
+      end
+
+      if !bookmark
+        # create one
+        now = Time.zone.now
+        new_bookmarks << {
+          user_id: tt.user_id,
+          topic_id: tt.topic_id,
+          post_id: tt.first_post_id,
+          reminder_at: tt.execute_at,
+          reminder_type: 6, # custom
+          created_at: now,
+          updated_at: now
+        }
+      else
+        if !bookmark.reminder_at
+          Bookmark.find(bookmark.id).update(reminder_at: tt.execute_at, reminder_type: 6)
```

Comment:
I can't seem to find the dev topic regarding this but we should not use any `ActiveRecord` methods in migrations. Everything has to be done via raw SQL.

Reply:
@tgxworld how do I still consistently forget this :sob: thanks for pointing that out.